rfcs/45-stageless.md

```diff
@@ -0,0 +1,996 @@
+When you supply a constant delta time value (the literal fixed timestep) inside the encapsulated systems, the result is consistent and repeatable behavior regardless of framerate (or even the presence of a GPU).
+
+```rust
+fn main() {
+    App::new()
+        .add_systems(
+            // Conveneniently ordering these systems relative to each other
+            chain!([
+               apply_forces,
+               apply_acceleration,
+               apply_velocity
+            ])
+            // Ensuring they run on a fixed timestep
+            // as part of the `CoreSchedule::FixedUpdate` schedule
+            .in_schedule(CoreSchedule::FixedUpdate)
+         )
+        .run();
+}
+```
+
+### States
+
+**State machines** allow for ad-hoc execution of complex logic in response to state transitions.
+They're often used for relatively high-level facets of app behavior like pausing, opening menus, and loading assets.
+State machines are stored as individual resources and their state types must implement the `State` trait.
+It's common to see enums used for state types.
+
+You can have multiple independently controlled states, or defined nested states with enums.
+The current state of type `S` can be read from the `CurrentState<S: State>` resource,
+while the upcoming state can be read from `NextState<S: State>`.
+
+Bevy provides a simple but convenient abstraction to link the transitions of a state `S::Variant` with run-on-demand schedules, specifically an `OnEnter(S::Variant)` and `OnExit(S::Variant)` schedules.
+An exclusive system called `apply_state_transition<S: State>` can be scheduled, which will retrieve and run these schedules as appropriate when a transition is queued in the `NextState<S: State>` resource.
+
+```rust
+fn main() {
+    App::new()
+        .add_state::<GameState>()
+        /* ... */
+        // These systems only run as part of the state transition logic
+        .add_system(load_map.in_schedule(OnEnter(GameState::Playing)))
+        // The .on_enter and .on_exit methods are equivalent to the pattern above
+        .add_system(autosave.on_exit(GameState::Playing))
+        // This system will be added under the `OnUpdate(GameState::Playing)` set
+        // Note that this is not a schedule:
+        // it is run as part of the main game schedule,
+        // but only runs if the current state is GameState::Playing
+        .add_system(run.in_set(OnUpdate(GameState::Playing)))
+        // This system will not be part of the set above, but will otherwise follow the same rules
+        .add_system(jump.run_if(state_equals(GameState::Playing)))
+        /* ... */
+        .run();
+}
+```
+
+By default, all state transitions are evaluated just before `CoreSystems::Update`, in a `CoreSystems::ApplyStateTransitions` set.
+Calling `App::add_state` adds the corresponding `CurrentState` and `NextState` resources, and registers a `apply_state_transition<S>` system.

```

Comment:
Probably worth calling out that (by default) this also adds the OnUpdate set, which is in the `CoreSystems::Update` set.